### PR TITLE
List migration logs

### PR DESCRIPTION
**Pull request by sandre1**
-  logs for list available (all), pending, applied migrations with applied-date visible
-  option --format FILE-FORMAT in CLI to list migrations (as logs)  in .edn and .json format

### CODE REVIEW

**Comment by yogthos** [comment]
Overall looks good to me, would just swap out cheshire for clojure.data.json

File: src/migratus/cli.clj [hunk 1]
```diff
@@ -95,17 +98,109 @@
 
       :else (no-match-message args summary))))
 
+(defn util-date-to-local-datetime [util-date]
+  (if (= util-date nil)
```
Comment (ieugen) [inline src/migratus/cli.clj:102]: Use a more readable way to test for nil: 
- https://clojuredocs.org/clojure.core/nil_q  
- https://clojuredocs.org/clojure.core/some_q 

I recommend: https://clojuredocs.org/clojure.core/when-some  .

File: src/migratus/cli.clj [hunk 2]
```diff
@@ -95,17 +98,109 @@
...
+    nil
+    (let [instant (.toInstant util-date)
+          zone-id (ZoneId/systemDefault)
+          local-datetime (.atZone instant zone-id)]
+      local-datetime)))
+
+(defn formatted-date [date]
+  (if (nil? date)
```
Comment (ieugen) [inline src/migratus/cli.clj:110]: See above - nil issue .

File: src/migratus/cli.clj [hunk 3]
```diff
@@ -95,17 +98,109 @@
...
+    nil
+    (core/format "%1tFT%<tTZ", date)))
+
+(defn parsed-migrations-data [cfg]
+  (let [all-migrations (migratus/all-migrations cfg)]
+    (map (fn [m] (let [{:keys [id name applied]} m
```
Comment (ieugen) [inline src/migratus/cli.clj:116]: try to avoid writing fn's inside map .
It makes the code harder to read. 

Well written code should read as prose (some say).

Define the function in let or using defn , with a name and use the name. 
The function should be named for what it does. 

If you can't find a name easy - it means the function is not that clear in what it does. 
Inline functions

File: src/migratus/cli.clj [hunk 4]
```diff
@@ -95,17 +98,109 @@
...
+                       to-local-date (if (nil? applied)
```
Comment (ieugen) [inline src/migratus/cli.clj:117]: nil again.

File: src/migratus/cli.clj [hunk 5]
```diff
@@ -95,17 +98,109 @@
...
+                                       nil (util-date-to-local-datetime applied))
+                       date-str (formatted-date to-local-date)]
+                   {:id id :name name :applied date-str})) all-migrations)))
+
+(defn pending-migrations [cfg]
+  (filter (fn [mig] (= nil (:applied mig))) (parsed-migrations-data cfg)))
```
Comment (ieugen) [inline src/migratus/cli.clj:123]: nil again

File: src/migratus/cli.clj [hunk 6]
```diff
@@ -95,17 +98,109 @@
...
+
+(defn applied-migrations [cfg]
+  (filter (fn [mig] (not= nil (:applied mig))) (parsed-migrations-data cfg)))
```
Comment (ieugen) [inline src/migratus/cli.clj:126]: nil again .

File: src/migratus/cli.clj [hunk 7]
```diff
@@ -95,17 +98,109 @@
...
+
+(defn cli-print-migrations-edn! [data]
+  (log/info data))
```
Comment (ieugen) [inline src/migratus/cli.clj:129]: one function to log it's argument? 
Why ?! 
O bit overkill.

File: src/migratus/cli.clj [hunk 8]
```diff
@@ -95,17 +98,109 @@
...
+
+(defn cli-print-migrations-json! [data]
+  (log/info (cheshire/generate-string data)))
+
+(defn cli-print-migrations! [data f]
```
Comment (ieugen) [inline src/migratus/cli.clj:134]: I think you can redo the whole printing of json and edn. 
You dance around the issue a lot with not much benefit. 

For output: 
You wither build the whole data once as a string and dump it, or you convert each row and print that. 

You can iterate through rows, you do that for plain output. 
You can do the same for edn / json . 

You can also use lazy evaluation in this case to make a list that renders it's row in the format of choosing.

File: src/migratus/cli.clj [hunk 9]
```diff
@@ -95,17 +98,109 @@
...
+
+
+(defn mig-print-fmt [data]
+  (log/info (table-line 67))
+  (log/info (core/format "%-18s%-25s%-22s%s",
+                         "| MIGRATION-ID" "| NAME" "| APPLIED" " |"))
+  (log/info (table-line 67))
+  (doall
+   (map
+    (fn [e]
```
Comment (ieugen) [inline src/migratus/cli.clj:157]: inline fn inside map !
See above!

File: src/migratus/cli.clj [hunk 10]
```diff
@@ -95,17 +98,109 @@
...
+  (let [str (str "%-" n "s")]
+    (core/format str, (col-width n))))
+
+
+(defn mig-print-fmt [data]
+  (log/info (table-line 67))
+  (log/info (core/format "%-18s%-25s%-22s%s",
+                         "| MIGRATION-ID" "| NAME" "| APPLIED" " |"))
+  (log/info (table-line 67))
+  (doall
```
Comment (ieugen) [inline src/migratus/cli.clj:155]: Use doseq instead of doall map .

doseq is used for
> Repeatedly executes body (presumably for side-effects) 
 https://clojuredocs.org/clojure.core/doseq

File: src/migratus/cli.clj [hunk 11]
```diff
@@ -95,17 +98,109 @@
...
+            fmt-str "| %1$-15s | %2$-22s | %3$-20s%4$s"]
+        (log/info (core/format fmt-str, id, name, applied? " |")))) data))
+  (log/info (table-line 67)))
+
+(defn pending-mig-print-fmt [data]
+  (log/info (table-line 43))
+  (log/info (core/format "%-17s%-24s%s",
+                         "| MIGRATION-ID" "| NAME" " |"))
+  (log/info (table-line 43))
+  (doall
```
Comment (ieugen) [inline src/migratus/cli.clj:171]: doseq + fn in map.

File: src/migratus/cli.clj [hunk 12]
```diff
@@ -95,17 +98,109 @@
...
+      (let [{:keys [id name]} e
+            fmt-str "| %1$-15s| %2$-22s%3$s"]
+        (log/info (core/format fmt-str, id, name, " |")))) data))
+  (log/info (table-line 43)))
+
 (defn run-list [cfg args]
-  (let [{:keys [options _arguments errors summary]} (parse-opts args list-cli-options :in-order true)]
+  (let [{:keys [options errors summary]} (parse-opts args list-cli-options :in-order true)
+        {:keys [available pending applied]} options
+        {f :format} options
```
Comment (ieugen) [inline src/migratus/cli.clj:182]: format should default to `plain` . 
Then you can move the if as a case inside a single fn that does the action. 

printing plain, json, edn is the same case. it's not a special case. 
You can process them in the same way IMO.

File: src/migratus/core.clj [hunk 13]
```diff
@@ -68,6 +68,25 @@
...
+  "Returns a list of all migrations from migration dir and db
+  with enriched data:
+    - date and time when was applied;
+    - description;"
+  [config store]
+  (let [completed-migrations (vec (proto/completed store))
+        available-migrations (mig/list-migrations config)
+        merged-migrations-data (apply merge completed-migrations available-migrations)
+        grouped-migrations-by-id (group-by :id merged-migrations-data)]
+    (map (fn [[_ v]] (apply merge v)) grouped-migrations-by-id)))
```
Comment (ieugen) [inline src/migratus/core.clj:81]: it would make the code clearer if you give the fn a name .

**Review by ieugen — CHANGES_REQUESTED** [review]

File: src/migratus/cli.clj (hunk 1 above)
Comment (sandre1) [reply src/migratus/cli.clj:102]: refactored if statement

File: src/migratus/cli.clj (hunk 2 above)
Comment (sandre1) [reply src/migratus/cli.clj:110]: refactored if-statement

File: src/migratus/cli.clj (hunk 3 above)
Comment (sandre1) [reply src/migratus/cli.clj:116]: named the anonymous fn

File: src/migratus/cli.clj (hunk 9 above)
Comment (sandre1) [reply src/migratus/cli.clj:157]: named the anonymous fn

File: src/migratus/core.clj (hunk 13 above)
Comment (sandre1) [reply src/migratus/core.clj:81]: named fn

File: src/migratus/cli.clj (hunk 10 above)
Comment (sandre1) [reply src/migratus/cli.clj:155]: changed (doall (map ...))  with doseq.

File: src/migratus/cli.clj (hunk 11 above)
Comment (sandre1) [reply src/migratus/cli.clj:171]: changed doall map with doseq
named the fn

File: src/migratus/cli.clj (hunk 12 above)
Comment (sandre1) [reply src/migratus/cli.clj:190]: Refactored code with plain text format as default format

**Comment by yogthos** [comment]
@ieugen thanks for doing a more thorough review, and I'm good to merge if everything looks good to you

File: src/migratus/cli.clj [hunk 14]
```diff
@@ -95,17 +101,99 @@
...
       :else (no-match-message args summary))))
 
+(defn util-date-to-local-datetime [util-date]
+  (when (some? util-date)
+    (let [instant (.toInstant util-date)
+          zone-id (ZoneId/systemDefault)
+          local-datetime (.atZone instant zone-id)]
+      local-datetime)))
+
+(defn formatted-date [date]
```
Comment (ieugen) [inline src/migratus/cli.clj:111]: If you format the date as string and print the string, then it's better to use date time formatters https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE_TIME

**Review by ieugen — APPROVED** [review]

**Comment by ieugen** [comment]
@yogthos : I think it's ready to merge. 
I am not particularly fond of terminal output with logging. 
It would be nice to improve that, but I think the code is an improvement to migratus and a step forward.

**Comment by yogthos** [comment]
Agreed, we could do a separate PR to make that nicer. And just released `1.5.3` to Clojars with the update @sandre1 thanks for spearheading the effort on this. :)
